src/oc_erchef/apps/oc_chef_wm/src/chef_wm_actions_queue_monitoring.erl

```diff
@@ -0,0 +1,473 @@
+                % calling process. This is recorded because checking the queue
+                % length and max is async via the worker_process, and the
+                % reply is sent back to this Pid via gen_server:reply()
+                sync_response_process = undefined
+               }).
+
+% exchange is /analytics, so encode the / as %2F
+% NOTE: oc_httpc client is configured to prepend /api
+-define(MAX_LENGTH_PATH, "/policies/%2Fanalytics/max_length").
+-define(QUEUE_LENGTH_PATH, "/queues/%2Fanalytics").
+-define(LOG_THRESHOLD, 0.8).
+-define(POOLNAME, rabbitmq_management_service).
+
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+-spec status() -> [{atom(), _}].
+status() ->
+    gen_server:call(?SERVER, status).
+
+% is current queue length >= max queue length?
+% Note, this simply checks the gen_server state, and does NOT
+% ping the RabbitMQ management console.
+% Pinging the management console is handled by an internal timer
+% so we never block waiting for an HTTP response.
+is_queue_at_capacity() ->
+    Timeout = envy:get(oc_chef_wm, rabbitmq_queue_length_timeout_millis, 5000, non_neg_integer),
+    try
+        gen_server:call(?SERVER, is_queue_at_capacity, Timeout)
+    catch Error:Reason ->
+        lager:warning("Queue monitor timeout ~p ~p ", [Error, Reason]),
+        true
+    end.
+
+% log that a message has been dropped (hasn't been sent to rabbitmq)
+% This is a cast so we don't wait for an `ok` response
+% back to the callee, which a call would have returned as well.
+message_dropped() ->
+    gen_server:cast(?SERVER, message_dropped).
+
+
+-spec stop() -> ok.
+stop() ->
+    gen_server:call(?SERVER, stop).
+
+
+%% Support functions --------------------------------------------
+% Synchronously update state by pinging the RabbitMQ Management Console
+sync_check_current_state() ->
+    gen_server:call(?SERVER, sync_check_current_state).
+
+% manually toggle the at capacity flag
+override_queue_at_capacity(AtCapacity) ->
+    gen_server:call(?SERVER, {override_queue_at_capacity, AtCapacity}).
+
+start_timer() ->
+    gen_server:call(?SERVER, start_timer).
+
+stop_timer() ->
+    gen_server:call(?SERVER, stop_timer).
+
+
+%% oc_httpc pool functions --------------------------------------------
+create_pool() ->
+    Pools = get_pool_configs(),
+    [oc_httpc:add_pool(PoolNameAtom, Config) || {PoolNameAtom, Config} <- Pools, Config /= []],
+    ok.
+
+delete_pool() ->
+    Pools = get_pool_configs(),
+    [ok = oc_httpc:delete_pool(PoolNameAtom) || {PoolNameAtom, _Config} <- Pools],
+    ok.
+
+get_pool_configs() ->
+    Config = envy:get(oc_chef_wm, ?POOLNAME, [], any),
+    [{?POOLNAME, Config}].
+
+%%-------------------------------------------------------------
+
+init([]) ->
+    % used to catch worker msgs
+    process_flag(trap_exit, true),
+    TRef = start_update_timer(),
+    {ok, #queue_monitor_state{timer=TRef, worker_process = undefined}}.
```

Comment:
As a sync request (see other comment  about splitting out some functions)  Basically it's a way to make sure erchef itself can't start handling requests until monitoring is in place - since the application complete its start until the supervisor reports success.